src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs

```diff
@@ -796,76 +842,127 @@ internal DefinitionResult GetDefinition(TextDocumentPosition textDocumentPositio
-                                schemaName);
-                        },
-                        timeoutOperation: (bindingContext) =>
+                            return result;
+                        }                       
+                    }
+                    catch (Exception ex)
+                    {
+                        // if any exceptions are raised return error result with message
+                        Logger.Write(LogLevel.Error, "Exception in GetDefinition " + ex.ToString());
+                        return new DefinitionResult
                         {
-                            // return error result
-                            return new DefinitionResult
-                            {
-                                IsErrorResult = true,
-                                Message = SR.PeekDefinitionTimedoutError,
-                                Locations = null
-                            };
-                        });
-
-                    // wait for the queue item
-                    queueItem.ItemProcessed.WaitOne();
-                    return queueItem.GetResultAsT<DefinitionResult>();
+                            IsErrorResult = true,
+                            Message = SR.PeekDefinitionError(ex.Message),
+                            Locations = null
+                        };
+                    }
+                    finally
+                    {
+                        Monitor.Exit(scriptParseInfo.BuildingMetadataLock);
+                    }
                 }
-                catch (Exception ex)
+                else
                 {
-                    // if any exceptions are raised return error result with message
-                    Logger.Write(LogLevel.Error, "Exception in GetDefinition " + ex.ToString());
+                    // User is not connected.
                     return new DefinitionResult
                     {
                         IsErrorResult = true,
-                        Message = SR.PeekDefinitionError(ex.Message),
+                        Message = SR.PeekDefinitionNotConnectedError,
                         Locations = null
                     };
                 }
-                finally
+            }
+
+            // then check the parents 
+            foreach (var i in selectedToken.Item2.ToList())
+            {
+                var token = selectedToken.Item2.Dequeue();
+                // Strip "[" and "]"(if present) from the token text to enable matching with the suggestions.
+                // The suggestion title does not contain any sql punctuation
+                string tokenText = TextUtilities.RemoveSquareBracketSyntax(token.Text);
+                textDocumentPosition.Position.Line = token.StartLocation.LineNumber;
+                textDocumentPosition.Position.Character = token.StartLocation.ColumnNumber;
+                if (scriptParseInfo.IsConnected && Monitor.TryEnter(scriptParseInfo.BuildingMetadataLock))
```

Comment:
Since this is the exact same logic, refactor to have 1 method, and just call with the relevant tokens